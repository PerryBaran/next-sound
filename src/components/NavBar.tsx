import Link from 'next/link';
import SearchBar from './SearchBar';

export default function NavBar() {

  return (
    <nav>
      <ul>
        <li>
          <Link href="/">Home</Link>
        </li>
        <li>
          <SearchBar />
        </li>
        <li>
          <Link href="/login">Login</Link>
        </li>
      </ul>
    </nav>
  )
}