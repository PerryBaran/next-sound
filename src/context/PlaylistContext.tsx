'use client'

import { createContext, useContext, useState } from 'react'

interface Song {
  image?: string
  artistName: string
  albumName: string
  songName: string
  audio: string
}

interface Playlist extends Song {
  key: string
}

const PlaylistContext = createContext({
  playlist: [] as Playlist[],
  playlistIndex: 0,
  skipSong: (i: number | boolean) => {},
  handleAddToPlaylist: (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {},
  removeFromPlaylist: (i: number) => {},
  playing: false,
  handlePlaying: (isPlaying: boolean | undefined) => {}
})

const usePlaylistContext = () => useContext(PlaylistContext)

function PlaylistProvider({ children }: { children: React.ReactNode }) {
  const [playlist, setPlaylist] = useState<Playlist[]>([])
  const [playlistIndex, setPlaylistIndex] = useState(0)
  const [playing, setPlaying] = useState(false)

  const addToPlaylist = (songs: Song[], addNext = false) => {
    const data = songs.map((song) => {
      return { ...song, key: crypto.randomUUID() }
    })

    setPlaylist((prev) => {
      if (addNext) {
        const clone = [...prev]
        clone.splice(playlistIndex + 1, 0, ...data)
        return clone
      }
      return [...prev, ...data]
    })
  }

  const handlePlayNow = () => {
    if (playlist.length !== 0) {
      setPlaylistIndex((prev) => prev + 1)
    }
    setPlaying(true)
  }

  const handleAddToPlaylist = (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {
    addToPlaylist(songs, addNext)
    if (playNow) {
      handlePlayNow()
    }
  }

  const removeFromPlaylist = (i: number) => {
    setPlaylist((prev) => {
      const clone = [...prev]
      clone.splice(i, 1)
      return clone
    })

    setPlaylistIndex((prev) => {
      if (i > prev || prev === 0) {
        return prev
      }
      return prev - 1
    })
  }

  const skipSong = (value: number | boolean) => {
    const lastPlaylistIndex = playlist.length - 1
    if (typeof value === 'number') {
      if (value > lastPlaylistIndex) {
        setPlaylistIndex(lastPlaylistIndex)
      } else if (value < 0) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex(value)
      }
    } else if (value) {
      if (playlistIndex === lastPlaylistIndex) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex((prev) => prev + 1)
      }
    } else {
      if (playlistIndex > 0) {
        setPlaylistIndex((prev) => prev - 1)
      } else {
        setPlaylistIndex(lastPlaylistIndex)
      }
    }
  }

  const handlePlaying = (isPlaying: boolean | undefined) => {
    if (isPlaying === undefined) {
      setPlaying((prev) => !prev)
    } else {
      setPlaying(isPlaying)
    }
  }

  return (
    <PlaylistContext.Provider
      value={{
        playlist,
        playlistIndex,
        handleAddToPlaylist,
        removeFromPlaylist,
        skipSong,
        playing,
        handlePlaying
      }}
    >
      {children}
    </PlaylistContext.Provider>
  )
}

export { usePlaylistContext, PlaylistProvider }
