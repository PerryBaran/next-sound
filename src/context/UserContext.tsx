'use client'

import { createContext, useContext, useState, useEffect } from 'react'
import { getUserById } from '@/requests/users'
import Cookie from 'js-cookie'
import jwtDecode from 'jwt-decode'

interface User {
  name?: string
  id?: string
}

const emptyUser = {
  name: '',
  id: ''
}

const UserContext = createContext({
  user: emptyUser,
  handleLogin: (data: User | undefined) => {}
})

const useUserContext = () => useContext(UserContext)

function UserProvider({ children }: { children: React.ReactNode }) {
  const [user, setUser] = useState(emptyUser)

  const handleLogin = (data: User | undefined) => {
    if (data === undefined) {
      Cookie.remove('userToken')
      setUser(emptyUser)
    } else {
      setUser((prev) => {
        return { ...prev, ...data }
      })
    }
  }

  useEffect(() => {
    ;(async () => {
      const token = Cookie.get('userToken')
      if (!token) return
      const { id }: { id: string } = jwtDecode(token)
      const response = await getUserById(id)
      handleLogin({
        name: response.name,
        id
      })
    })()
  }, [])

  return (
    <UserContext.Provider value={{ user, handleLogin }}>
      {children}
    </UserContext.Provider>
  )
}

export { useUserContext, UserProvider }
