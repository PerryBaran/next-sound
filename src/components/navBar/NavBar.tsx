"use client"

import Link from "next/link"
import { useRouter } from "next/navigation"
import { usePathname } from "next/navigation"
import { useRef } from "react"
import { useUserContext } from "@/context/UserContext"
import css from "./navbar.module.css"
import { home, login, logout, profile, search, upload } from "@/media/icons"
import Image from "next/image"

export default function NavBar() {
  const router = useRouter()
  const {
    user: { name }
  } = useUserContext()
  const pathname = usePathname()
  const searchRef = useRef<HTMLInputElement>(null)

  const submitHandler = (e: React.FormEvent) => {
    e.preventDefault()
    const query = searchRef?.current?.value
    if (query) {
      router.push(`search/${query}`)
    }
  }

  const classes = (linkPath: string) => {
    return `${pathname === linkPath ? css.active : ""} ${css.link}`
  }

  return (
    <nav className={css.container}>
      <ul>
        <li>
          <Link href="/">
            <h1>Next-Sound</h1>
          </Link>
        </li>
        <li className={classes("/")}>
          <Link href="/">
            <Image src={home} alt="home" height={30} width={30} />
            <span>Home</span>
          </Link>
        </li>
        <li>
          <form onSubmit={submitHandler} className={css.search}>
            <input type="text" ref={searchRef} placeholder="Search" />
            <button type="submit">
              <Image src={search} alt="search" height={15} width={15} />
            </button>
          </form>
        </li>
        {name ? (
          <>
            <li className={classes(`/profile/${name}`)}>
              <Link href={`/profile/${name}`}>
                <Image src={profile} alt="Profile" height={30} width={30} />
                <span>Profile</span>
              </Link>
            </li>
            <li className={classes("/upload")}>
              <Link href="/upload">
                <Image src={upload} alt="upload" height={30} width={30} />
                <span>Upload</span>
              </Link>
            </li>
            <li className={classes("/logout")}>
              <Link href="/logout">
                <Image src={logout} alt="logout" height={30} width={30} />
                <span>Logout</span>
              </Link>
            </li>
          </>
        ) : (
          <li className={classes("/login")}>
            <Link href="/login">
              <Image src={login} alt="login" height={30} width={30} />
              <span>Login</span>
            </Link>
          </li>
        )}
      </ul>
    </nav>
  )
}
