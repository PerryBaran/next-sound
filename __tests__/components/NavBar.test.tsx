import { fireEvent, render, screen } from "@testing-library/react"
import { createContext, useContext } from "react"
import * as navigation from "next/navigation"
import NavBar from "../../src/components/navBar/NavBar"
import * as userContext from "../../src/context/UserContext"

describe("NavBar", () => {
  let mockNav = {
    back: jest.fn(),
    forward: jest.fn(),
    reset: jest.fn(),
    push: jest.fn(),
    pull: jest.fn(),
    refresh: jest.fn(),
    replace: jest.fn(),
    prefetch: jest.fn()
  }

  beforeEach(() => {
    mockNav = {
      back: jest.fn(),
      forward: jest.fn(),
      reset: jest.fn(),
      push: jest.fn(),
      pull: jest.fn(),
      refresh: jest.fn(),
      replace: jest.fn(),
      prefetch: jest.fn()
    }

    jest.spyOn(navigation, "useRouter").mockImplementation(() => mockNav)
  })

  describe("user logged out", () => {
    beforeEach(() => {
      const mockContext = createContext({
        user: {
          id: "",
          name: ""
        },
        handleLogin: () => {}
      })

      jest
        .spyOn(userContext, "useUserContext")
        .mockImplementation(() => useContext(mockContext))
    })

    test("renders correctly", () => {
      render(<NavBar />)

      expect(screen.getByRole("navigation")).toBeInTheDocument()
      expect(screen.getByRole("list")).toBeInTheDocument()
      expect(screen.getAllByRole("listitem")).toHaveLength(4)

      const heading = screen.getByRole("heading")
      expect(heading).toHaveTextContent(/next-sound/i)
      expect(heading.parentElement).toHaveAttribute("href", "/")

      const home = screen.getByText(/home/i)
      expect(home.parentElement).toHaveAttribute("href", "/")
      expect(home.previousSibling).toHaveAttribute("alt", "home")

      const formButton = screen.getByRole("button")
      expect(screen.getByRole("form")).toBeInTheDocument()
      expect(screen.getByPlaceholderText(/search/i)).toHaveAttribute(
        "type",
        "text"
      )
      expect(formButton).toHaveAttribute("type", "submit")
      expect(formButton.firstChild).toHaveAttribute("alt", "search")

      const login = screen.getByText(/login/i)
      expect(login.parentElement).toHaveAttribute("href", "/login")
      expect(login.previousSibling).toHaveAttribute("alt", "login")
    })

    test("search functionality", () => {
      render(<NavBar />)

      expect(mockNav.push).toBeCalledTimes(0)

      const search = "search text"
      fireEvent.change(screen.getByPlaceholderText(/search/i), {
        target: { value: search }
      })
      fireEvent.click(screen.getByRole("button"))

      expect(mockNav.push).toBeCalledTimes(1)
      expect(mockNav.push).toBeCalledWith(`search/${search}`)
    })

    test("snapshot", () => {
      const { asFragment } = render(<NavBar />)

      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe("user logged in", () => {
    const userName = "test-name"

    beforeEach(() => {
      const mockContext = createContext({
        user: {
          id: "test-id",
          name: userName
        },
        handleLogin: () => {}
      })

      jest
        .spyOn(userContext, "useUserContext")
        .mockImplementation(() => useContext(mockContext))
    })

    test("renders correctly", () => {
      render(<NavBar />)

      expect(screen.getByRole("navigation")).toBeInTheDocument()
      expect(screen.getByRole("list")).toBeInTheDocument()
      expect(screen.getAllByRole("listitem")).toHaveLength(6)

      const heading = screen.getByRole("heading")
      expect(heading).toHaveTextContent(/next-sound/i)
      expect(heading.parentElement).toHaveAttribute("href", "/")

      const home = screen.getByText(/home/i)
      expect(home.parentElement).toHaveAttribute("href", "/")
      expect(home.previousSibling).toHaveAttribute("alt", "home")

      const formButton = screen.getByRole("button")
      expect(screen.getByRole("form")).toBeInTheDocument()
      expect(screen.getByPlaceholderText(/search/i)).toHaveAttribute(
        "type",
        "text"
      )
      expect(formButton).toHaveAttribute("type", "submit")
      expect(formButton.firstChild).toHaveAttribute("alt", "search")

      const profile = screen.getByText(/profile/i)
      expect(profile.parentElement).toHaveAttribute(
        "href",
        `/profile/${userName}`
      )
      expect(profile.previousSibling).toHaveAttribute("alt", "profile")

      const upload = screen.getByText(/upload/i)
      expect(upload.parentElement).toHaveAttribute("href", "/upload")
      expect(upload.previousSibling).toHaveAttribute("alt", "upload")

      const logout = screen.getByText(/logout/i)
      expect(logout.parentElement).toHaveAttribute("href", "/logout")
      expect(logout.previousSibling).toHaveAttribute("alt", "logout")
    })

    test("snapshot", () => {
      const { asFragment } = render(<NavBar />)

      expect(asFragment()).toMatchSnapshot()
    })
  })
})
