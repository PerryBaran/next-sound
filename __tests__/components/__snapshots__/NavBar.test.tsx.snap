// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar user logged in snapshot 1`] = `
<DocumentFragment>
  <nav
    class="container"
  >
    <ul>
      <li>
        <a
          href="/"
        >
          <h1>
            Next-Sound
          </h1>
        </a>
      </li>
      <li
        class=" link"
      >
        <a
          href="/"
        >
          <img
            alt="home"
            data-nimg="1"
            decoding="async"
            height="30"
            loading="lazy"
            src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
            srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
            style="color: transparent;"
            width="30"
          />
          <span>
            Home
          </span>
        </a>
      </li>
      <li>
        <form
          aria-label="form"
          class="search"
        >
          <input
            placeholder="Search"
            type="text"
          />
          <button
            type="submit"
          >
            <img
              alt="search"
              data-nimg="1"
              decoding="async"
              height="15"
              loading="lazy"
              src="/_next/image?url=%2Fimg.jpg&w=32&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=16&q=75 1x, /_next/image?url=%2Fimg.jpg&w=32&q=75 2x"
              style="color: transparent;"
              width="15"
            />
          </button>
        </form>
      </li>
      <li
        class=" link"
      >
        <a
          href="/profile/test-name"
        >
          <img
            alt="profile"
            data-nimg="1"
            decoding="async"
            height="30"
            loading="lazy"
            src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
            srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
            style="color: transparent;"
            width="30"
          />
          <span>
            Profile
          </span>
        </a>
      </li>
      <li
        class=" link"
      >
        <a
          href="/upload"
        >
          <img
            alt="upload"
            data-nimg="1"
            decoding="async"
            height="30"
            loading="lazy"
            src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
            srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
            style="color: transparent;"
            width="30"
          />
          <span>
            Upload
          </span>
        </a>
      </li>
      <li
        class=" link"
      >
        <a
          href="/logout"
        >
          <img
            alt="logout"
            data-nimg="1"
            decoding="async"
            height="30"
            loading="lazy"
            src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
            srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
            style="color: transparent;"
            width="30"
          />
          <span>
            Logout
          </span>
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;

exports[`NavBar user logged out snapshot 1`] = `
<DocumentFragment>
  <nav
    class="container"
  >
    <ul>
      <li>
        <a
          href="/"
        >
          <h1>
            Next-Sound
          </h1>
        </a>
      </li>
      <li
        class=" link"
      >
        <a
          href="/"
        >
          <img
            alt="home"
            data-nimg="1"
            decoding="async"
            height="30"
            loading="lazy"
            src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
            srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
            style="color: transparent;"
            width="30"
          />
          <span>
            Home
          </span>
        </a>
      </li>
      <li>
        <form
          aria-label="form"
          class="search"
        >
          <input
            placeholder="Search"
            type="text"
          />
          <button
            type="submit"
          >
            <img
              alt="search"
              data-nimg="1"
              decoding="async"
              height="15"
              loading="lazy"
              src="/_next/image?url=%2Fimg.jpg&w=32&q=75"
              srcset="/_next/image?url=%2Fimg.jpg&w=16&q=75 1x, /_next/image?url=%2Fimg.jpg&w=32&q=75 2x"
              style="color: transparent;"
              width="15"
            />
          </button>
        </form>
      </li>
      <li
        class=" link"
      >
        <a
          href="/login"
        >
          <img
            alt="login"
            data-nimg="1"
            decoding="async"
            height="30"
            loading="lazy"
            src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
            srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
            style="color: transparent;"
            width="30"
          />
          <span>
            Login
          </span>
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
